source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -486,26 +503,37 @@ absl::optional<QueuedChunkPtr> ChunkQueue::pop(bool undelivered_only) {
   }
   QueuedChunkPtr chunk = std::move(queue_.front());
   queue_.pop_front();
-  bytes_enqueued_ -= chunk->data.length();
-  return chunk;
-}
+  bytes_enqueued_ -= chunk->buffer_length;
 
-const QueuedChunk& ChunkQueue::consolidate(bool delivered) {
-  if (queue_.size() == 1) {
-    queue_.front()->delivered = delivered;
-    return *(queue_.front());
-  }
+  // Copy the data out.
```

Comment:
I think you are trading a problem with fragmentation of the buffer to an equally bad problem of incurring multiple copies. This code adds two memory copies, instead of 0 previously.

Reply:
Yeah, that's right.

So, the goal of this PR is to avoid have one buffer for each body chunk and stored in the chunk queue as here: https://github.com/envoyproxy/envoy/blob/9c1dcef98fe3982e26ec9e24ccd19de39862949d/source/extensions/filters/http/ext_proc/processor_state.h#L32C26-L32C26

Since each buffer is 4kB minimal, and if received body chunk is very small, this will waste memory.  So, the idea is to maintain one buffer for all body chunks received, and maintain a logical view of the chunk in the queue.  That avoid memory wasting when the data is pending to be processed while waiting in the queue.  However, the memory  copy and fragmentation does happen when the data is copied out and shipped.  

@KBaichoo  do you have some good ideas?

Reply:
Adding @tyxia @stevenzzzz 

BTW, the reason we are copying the buffer out is due to the requirement to send out each chunk after body mutation as here: https://github.com/envoyproxy/envoy/blob/f9ffd76e1b6982666d2ac0556fde441357d25341/source/extensions/filters/http/ext_proc/processor_state.cc#L311

Reply:
Think we should discuss this offline where we have more bandwidth.

Reply:
Actually, there is a solution to address @yanavlasov 's comments.  We can just using move() methods to move certain length of data out of the received buffer for mutation. This optimize things. 

Also, using a consolidate buffer from beginning, and just maintain a logical view of each streamed body has the following benefits:

1) It avoids  the original problem that maintaining a copy of such buffer for each streamed body in the queue 
https://github.com/envoyproxy/envoy/blob/f17f1fdd644cc4b32f65579184757e3975843db6/source/extensions/filters/http/ext_proc/processor_state.h#L32

Note, Envoy minimum page size is 4k.

https://github.com/envoyproxy/envoy/blob/f17f1fdd644cc4b32f65579184757e3975843db6/source/common/buffer/buffer_impl.h#L358

So, have a copy of each streamed body in the queue
https://github.com/envoyproxy/envoy/blob/f17f1fdd644cc4b32f65579184757e3975843db6/source/extensions/filters/http/ext_proc/processor_state.h#L208

 is wasting memory when the body size is small, like 1 byte. It's basically 1/4096 memory utilization efficiency for each queue element.

2) In BUFFERED_PARTIAL mode,  the buffer needs to be consolidated before sending to ext_proc server for processing.  With the buffer is already consolidated, such buffer consolidation process can be avoided.